### PR TITLE
Add `hash_engine` as binary dependency to `orchestrator`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Use the new [cargo binary dependency feature](https://rust-lang.github.io/rfcs/3028-cargo-binary-dependencies.html) to build the `hash_engine` binary as a dependency.

## ⚠️ Known issues

The `bindeps` feature is currently unstable, so we either have to pass `-Z bindeps` as an argument to cargo or specify it in a [configuration file](https://doc.rust-lang.org/cargo/reference/config.html). Because the former way is distracting, I chose to add a `.cargo/config.toml`. Users, which have this file locally may need to adjust it manually or move their config-file to another location (for example to the home folder).

## 🐾  Next steps

- When merging #229, we either need to adjust the CI again to build python correctly, or (better)
- Move the Python build-logic into `build.rs` (make python a required dependency)

## 🔍 What does this change?

- Update the toolchain to the latest Rust version to get the latest cargo version
- Specify the `hash_engine` binary as a dependency for the orchestrator
- Adjust the orchestrator to use the artifact
- Remove binary crates from default members as they are now built on demand
- Update the CI to build the tests directly instead of the dependencies

### 📜 Does this require a change to the docs?

No, the README doesn't currently say that we explicitly needed to build the binary for the engine to run. 

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201793759102797/1201481007343181/f) (_internal_)

## 🛡 What tests cover this?

The CI is covering these changes

## ❓ How to test this?

1. Clean the project
1. Run the CLI or test without building the project
1. Change a file inside the project
1. goto 2.

## 📹 Demo

After changing a source file, the `hash_engine` binary is built:
```
❯ cargo test
   Compiling hash_engine_lib v0.0.0 (/home/ghost/Code/hash/packages/engine)
   Compiling hash_engine v0.0.0 (/home/ghost/Code/hash/packages/engine/bin/hash_engine)
   Compiling orchestrator v0.0.0 (/home/ghost/Code/hash/packages/engine/lib/orchestrator)
    Finished test [unoptimized + debuginfo] target(s) in 17.02s

```

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -254,9 +254,8 @@ jobs:
           cd ..
           rm -rf tmp # Delete the tmp folder
 
-      - name: Build test dependencies
-        if: matrix.directory == 'packages/engine'
-        run: cargo build -p hash_engine --manifest-path ${{ matrix.directory }}/Cargo.toml ${{ matrix.flags }}
+      - name: Build tests
```
Comment (Alfred-Mountfield) [inline .github/workflows/rust.yml:257]: Why is this necessary? Can't we just run tests now?

File: packages/engine/Cargo.toml [hunk 2]
```diff
@@ -5,7 +5,7 @@ edition = "2021"
 
 [workspace]
 members = [".", "bin/*", "lib/*"]
-default-members = [".", "bin/*", "lib/*"]
+default-members = [".", "lib/*"]
```
Comment (Alfred-Mountfield) [inline packages/engine/Cargo.toml:8]: I know it functionally doesn't have a difference but I'm not sure we want to make this change. Conceptually we want to build the libraries and binaries by default. It's actually quite weird that the libraries force the binaries to be rebuilt, but it's what we've wanted for a while so I'm not complaining about that.

I just think that if someone looks at default-members they should see we're building all of them by default.

File: packages/engine/lib/orchestrator/src/process/local.rs [hunk 3]
```diff
@@ -10,11 +10,7 @@ use hash_engine_lib::{
 
 use crate::process;
 
-#[cfg(debug_assertions)]
-const PROCESS_PATH_DEFAULT: &str = "./target/debug/hash_engine";
-
-#[cfg(not(debug_assertions))]
-const PROCESS_PATH_DEFAULT: &str = "./target/release/hash_engine";
+const PROCESS_PATH_DEFAULT: &str = env!("CARGO_BIN_FILE_HASH_ENGINE");
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/orchestrator/src/process/local.rs:13]: Is this equivalent? What happens if you're not running through cargo

File: packages/engine/lib/orchestrator/Cargo.toml [hunk 4]
```diff
@@ -20,4 +21,5 @@ uuid = { version = "0.8.1", features = ["v4", "serde"] }
 
 [features]
 default = ["build-nng"]
-build-nng = ["hash_engine_lib/build-nng"]
+build-nng = ["hash_engine_lib/build-nng", "hash_engine/build-nng"]
+texray = ["hash_engine/texray"]
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/orchestrator/Cargo.toml:25]: Are these features necessary now just so they can be forwarded to the dependency?

File: .cargo/config.toml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+[unstable]
+bindeps = true
```
Comment (Alfred-Mountfield) [inline .cargo/config.toml:2]: I don't really like us adding this file as it messes up local configs for devs. Perhaps we can commit it, and also add it to .gitignore so people can have their own local configs.

It makes it a little hard to specifically commit changes to the repo but I think it might be the best compromise.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (TimDiekmann) [reply .github/workflows/rust.yml:257]: It's not needed, but when adding the python we need to build the cdylib to link them. Also when looking at the CI, it's nice, if the jobs are separated.

File: packages/engine/lib/orchestrator/src/process/local.rs (hunk 3 above)
Comment (TimDiekmann) [reply packages/engine/lib/orchestrator/src/process/local.rs:13]: The artifact is built as a dependency, e.g. `target/debug/deps/artifact/hash_engine-abcdeadbeef`. We could use the old path as fallback.

File: packages/engine/Cargo.toml (hunk 2 above)
Comment (TimDiekmann) [reply packages/engine/Cargo.toml:8]: Good point.

File: .cargo/config.toml (hunk 5 above)
Comment (TimDiekmann) [reply .cargo/config.toml:2]: Yeah, that is unfortunate, thus I added this to the Known-issues-section. I put it in the project root for the CI to pick it up, as it is running the jobs from the project root. When running the command from `packages/engine`, the config will be merged as specified in [the cargo documentation](https://doc.rust-lang.org/cargo/reference/config.html). Thus, you could keep your file in `packages/engine/.cargo/config.toml` and it will be merged with the provided one.

I don't like to commit an ignored file, we could simply not commit it at all and mention it in the docs, that the flag has to be passed or the config has to be generated.

File: packages/engine/lib/orchestrator/Cargo.toml (hunk 4 above)
Comment (TimDiekmann) [reply packages/engine/lib/orchestrator/Cargo.toml:25]: Cargo just looks up all features requested for a specific crate (i.e. `hash_engine_lib`) and combines them. Thus those features are not needed (actually no features are required for the orchestrator). I adjusted it accordingly.

File: packages/engine/lib/orchestrator/src/process/local.rs (hunk 3 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/orchestrator/src/process/local.rs:13]: Are we not able to generate a clean `hash_engine` file anymore then?

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Alfred-Mountfield) [reply .github/workflows/rust.yml:257]: > It's not needed, but when adding the python we need to build

I'm confused, is it needed or is it not?

If it's not needed then I don't think we should have it

File: packages/engine/lib/orchestrator/src/process/local.rs (hunk 3 above)
Comment (TimDiekmann) [reply packages/engine/lib/orchestrator/src/process/local.rs:13]: I added a commit to use `target/<debug|release>/hash_engine` if the artifact was not generated.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (TimDiekmann) [reply .github/workflows/rust.yml:257]: It's not needed currently.

File: packages/engine/lib/orchestrator/src/process/local.rs [hunk 6]
```diff
@@ -10,11 +10,13 @@ use hash_engine_lib::{
 
 use crate::process;
 
+const PROCESS_PATH_DEFAULT: &str = env!("CARGO_BIN_FILE_HASH_ENGINE");
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/orchestrator/src/process/local.rs:13]: Can we rename these to `ENGINE_PATH_DEFAULT` or `ENGINE_BIN_PATH_...` 

Github won't let me suggest changes

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Alfred-Mountfield) [reply .github/workflows/rust.yml]: Sorry I just understood what you meant, when we add the Python changes that are currently WIP, we'll need to run a build as a separate step so we can link the cdylib in between building and running the tests. Gotcha!

Still think removing it for now was the right idea :) We might want to do the Python linking in build.rs or something later

**Review by Alfred-Mountfield — APPROVED** [review]
